modules/pytket-iqm/pytket/extensions/iqm/backends/iqm.py

```diff
@@ -0,0 +1,344 @@
+        """
+        circuits = list(circuits)
+        n_shots_list = Backend._get_n_shots_as_list(
+            n_shots,
+            len(circuits),
+            optional=False,
+        )
+
+        if valid_check:
+            self._check_all_circuits(circuits)
+
+        postprocess = kwargs.get("postprocess", False)
+
+        handles = []
+        for i, (c, n_shots) in enumerate(zip(circuits, n_shots_list)):
+            if postprocess:
+                c0, ppcirc = prepare_circuit(c, allow_classical=False, xcirc=_xcirc)
+                ppcirc_rep = ppcirc.to_dict()
+            else:
+                c0, ppcirc_rep = c, None
+            instrs = _translate_iqm(c0)
+            qm = [
+                SingleQubitMapping(logical_name=str(qb), physical_name=_as_name(qb))
+                for qb in c.qubits
+            ]
+            iqmc = IQMCircuit(
+                name=c.name if c.name else f"circuit_{i}", instructions=instrs
+            )
+            run_id = self._client.submit_circuit(iqmc, qm, shots=n_shots)
+            handles.append(ResultHandle(run_id.bytes, json.dumps(ppcirc_rep)))
+        for handle in handles:
+            self._cache[handle] = dict()
+        return handles
+
+    def _update_cache_result(
+        self, handle: ResultHandle, result_dict: Dict[str, BackendResult]
+    ) -> None:
+        if handle in self._cache:
+            self._cache[handle].update(result_dict)
+        else:
+            self._cache[handle] = result_dict
+
+    def circuit_status(self, handle: ResultHandle) -> CircuitStatus:
+        self._check_handle_type(handle)
+        run_id = UUID(bytes=cast(bytes, handle[0]))
+        run_result = self._client.get_run(run_id)
+        status = run_result.status
+        if status is RunStatus.PENDING:
+            return CircuitStatus(StatusEnum.SUBMITTED)
+        elif status is RunStatus.READY:
+            measurements = run_result.measurements
+            shots = OutcomeArray.from_readouts(
+                np.array(
+                    [[r[0] for r in rlist] for cbstr, rlist in measurements.items()],
+                    dtype=int,
+                )
+                .transpose()
+                .tolist()
+            )
+            ppcirc_rep = json.loads(cast(str, handle[1]))
+            ppcirc = Circuit.from_dict(ppcirc_rep) if ppcirc_rep is not None else None
+            self._update_cache_result(
+                handle, {"result": BackendResult(shots=shots, ppcirc=ppcirc)}
+            )
+            return CircuitStatus(StatusEnum.COMPLETED)
+        else:
+            assert status is RunStatus.FAILED
+            return CircuitStatus(StatusEnum.ERROR, run_result.message)
+
+    def get_result(self, handle: ResultHandle, **kwargs: KwargTypes) -> BackendResult:
+        """
+        See :py:meth:`pytket.backends.Backend.get_result`.
+        Supported kwargs: `timeout` (default 900).
+        """
+        try:
+            return super().get_result(handle)
```

Comment:
I am not sure that I have full understood why this is not handle as the return in line 283?

Reply:
This just checks if it is in the cache, returns it if so, and otherwise raises `CircuitNotRunError`. It's a bit of a funny pattern, I agree, but seems to be followed elsewhere, e.g. in the `aqt` extension.